### PR TITLE
Add rate limiter

### PR DESCRIPTION
**Pull request by eyurtsev**
This PR adds a simple rate limiter based on a token bucket.

I would love to extend RunnableBinding with this, we just need to make
sure there's no funny async vs. threading business.

This should be sufficient for benchmarking for now.

### CODE REVIEW

File: langchain_benchmarks/rate_limiting.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        Args:
+            requests_per_second: The number of tokens to add per second to the bucket.
+                Must be at least 1.
+            tokens_per_request: number of tokens that a request "costs".
+                This is 1 by default.
+            check_every_n_seconds: check whether the tokens are available
+                every this many seconds. Can be a float to represent
+                fractions of a second.
+        """
+        if requests_per_second < 1:
```
Comment (hinthornw) [inline langchain_benchmarks/rate_limiting.py:36]: Why?

File: langchain_benchmarks/tool_usage/agents.py [hunk 2]
```diff
@@ -24,33 +25,44 @@ def _ensure_output_exists(inputs: dict) -> dict:
 
 class OpenAIAgentFactory:
     def __init__(
-        self, task: ToolUsageTask, *, model: str = "gpt-3.5-turbo-16k"
+        self,
+        task: ToolUsageTask,
+        *,
+        model: str = "gpt-3.5-turbo-16k",
```
Comment (hinthornw) [inline langchain_benchmarks/tool_usage/agents.py:31]: ooc: why do we need a default here/

File: langchain_benchmarks/rate_limiting.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            if self.last is None:
+                self.last = now
+
+            elapsed = now - self.last
+
+            if elapsed * self.requests_per_second > 1:
+                self.available_tokens += elapsed * self.requests_per_second
+                self.last = now
+
+            self.available_tokens = min(self.available_tokens, self.requests_per_second)
```
Comment (hinthornw) [inline langchain_benchmarks/rate_limiting.py:71]: For noobs, could we add a comment saying something like "requests_per_second is also the maximum bucket size?

File: langchain_benchmarks/rate_limiting.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            requests_per_second: The number of tokens to add per second to the bucket.
+                Must be at least 1.
+            tokens_per_request: number of tokens that a request "costs".
+                This is 1 by default.
+            check_every_n_seconds: check whether the tokens are available
+                every this many seconds. Can be a float to represent
+                fractions of a second.
+        """
+        if requests_per_second < 1:
+            raise ValueError("Rate must be at least 1 request per second")
```
Comment (hinthornw) [inline langchain_benchmarks/rate_limiting.py:37]: For the lazy user who doesn't read the docstring, thoughts on adding a more instructional recommendation here? I naively would assume I could let there be 1/4 requests per second, for instance

File: langchain_benchmarks/rate_limiting.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+from langchain.schema.runnable import Runnable, RunnableLambda
+from langchain.schema.runnable.utils import Input, Output
+
+
+class RateLimiter:
+    def __init__(
+        self,
+        *,
+        requests_per_second: float = 1,
+        tokens_per_request: int = 1,
```
Comment (hinthornw) [inline langchain_benchmarks/rate_limiting.py:15]: Given our domain, could we clarify that tokens are NOT LLM tokens? I could imagine someone thinking this being the number of ~words we can send to the LLM per request. Maybe link to the token bucket algorithm for complete novices as well.

This is especially the case given that some providers rate limit based on e.g., prompt tokens or total tokens per minute/hour

File: langchain_benchmarks/rate_limiting.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        # A lock to ensure that tokens can only be consumed by one thread
+        # at a given time.
+        self._consume_lock = threading.Lock()
+        # tokens per request sets how many tokens
+        self.tokens_per_request = tokens_per_request
+        # The last time we tried to consume tokens.
+        self.last: Optional[time.time] = None
+        self.check_every_n_seconds = check_every_n_seconds
+
+    def consume(self) -> bool:
```
Comment (hinthornw) [inline langchain_benchmarks/rate_limiting.py:50]: Should this be private?

File: langchain_benchmarks/rate_limiting.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        if requests_per_second < 1:
+            raise ValueError("Rate must be at least 1 request per second")
+        self.requests_per_second = requests_per_second
+        # Number of tokens in the bucket.
+        self.available_tokens = 0.0
+        # A lock to ensure that tokens can only be consumed by one thread
+        # at a given time.
+        self._consume_lock = threading.Lock()
+        # tokens per request sets how many tokens
+        self.tokens_per_request = tokens_per_request
```
Comment (hinthornw) [inline langchain_benchmarks/rate_limiting.py:45]: Do we also need to check that `tokens_per_request <= requests_per_second` ? 

Given that we are treating `requests_per_second` as the max bucket size, we'd never consume tokens if that condition doesn't hold, right?

File: langchain_benchmarks/rate_limiting.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        A runnable lambda that acts as a throttled passthrough.
+    """
+
+    def _rate_limited_passthrough(input: dict, **kwargs: Any) -> dict:
+        """Throttle the input."""
+        rate_limiter.wait()
+        return input
+
+    return (
+        RunnableLambda(_rate_limited_passthrough).with_config({"name": "Rate Limit"})
```
Comment (hinthornw) [inline langchain_benchmarks/rate_limiting.py:105]: You could also just have the function name and docstring be prettier to make the default name nice. The function is rendered in the metadata

File: langchain_benchmarks/rate_limiting.py (hunk 5 above)
Comment (hinthornw) [reply langchain_benchmarks/rate_limiting.py:15]: Similarly, should call `requests_per_second` something like `tokens_per_second` to make it more clear?

**Review by hinthornw — COMMENTED** [review]
Some questions/comments mainly around docs and communication. The algorithm looks fine to me. May be good to add tests

File: langchain_benchmarks/tool_usage/agents.py (hunk 2 above)
Comment (eyurtsev) [reply langchain_benchmarks/tool_usage/agents.py:31]: I can remove in a separate PR -- probably shouldn't be here

File: langchain_benchmarks/rate_limiting.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            if self.available_tokens >= self.tokens_per_request:
+                self.available_tokens -= self.tokens_per_request
+                return True
+
+            return False
+
+    def wait(self) -> None:
+        """Blocking call to wait until the given number of tokens are available."""
+        while not self.consume():
+            time.sleep(self.check_every_n_seconds)
```
Comment (akira) [inline langchain_benchmarks/rate_limiting.py:82]: nit, may be able to use `threading.Condition` with wait / notify if wanting to avoid polling.

Comment (eyurtsev) [reply langchain_benchmarks/rate_limiting.py:87]: I don't have much experience with threading primitives. Any suggestions on how to re-write with condition? My understanding is that something will still need to do notification in this case, so at least one thread will need to do polling. is that right or is there a good way to re-write this?

Comment (akira) [reply langchain_benchmarks/rate_limiting.py:87]: Ah yes, that's correct - misread this and thought I saw a thread, but I see that it's being updated inline.  As you mentioned, the way to do that would be to create a background thread that replenishes the tokens and notifies the other threads waiting this that condition. But given we don't have that, would prob leave as is.

**Pull request merged**